### PR TITLE
Npm1300 led driver mfd

### PR DESCRIPTION
**Pull request by aasinclair**

### CODE REVIEW

**Comment by simonguinot** [comment]
Hi @aasinclair. Thanks for this driver. I'll review it later this week (or this week-end in the worse case).

File: drivers/led/led_npm1300.c [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+{
+	const struct led_npm1300_config *config = dev->config;
+	int led;
+	int ret;
+
+	if (!device_is_ready(config->mfd)) {
+		return -ENODEV;
+	}
+
+	for (led = 0; led < NPM1300_LED_PINS; led++) {
```
Comment (gmarull) [inline drivers/led/led_npm1300.c:78]: `uint8_t led = 0U;`

File: drivers/led/led_npm1300.c [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+		if (config->mode[led] <= NPM1300_LED_HOST) {
+			ret = mfd_npm1300_reg_write(config->mfd, NPM_LED_BASE,
```
Comment (gmarull) [inline drivers/led/led_npm1300.c:80]: ret can be scoped at this level

File: dts/bindings/led/nordic,npm1300-led.yaml [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+  The nPM1300 has three LED outputs.
+  Each LED can automatically display error or charging status,
+  or be controlled by software.
+
+compatible: "nordic,npm1300-led"
+
+properties:
+  nordic,led0-mode:
+    type: int
+    description: LED 0 mode
```
Comment (gmarull) [inline dts/bindings/led/nordic,npm1300-led.yaml:16]: suggestion here: maybe use a string enum, and use the ENUM_IDX() macros?

e.g. `nordic,led0-mode = "charging";`

File: drivers/led/led_npm1300.c (hunk 2 above)
Comment (simonguinot) [reply drivers/led/led_npm1300.c:80]: In addition `ret` is not used at all in this function. Please return an error if the initialization fails.

**Review by gmarull — DISMISSED** [review]

**Review by fabiobaltieri — DISMISSED** [review]

**Review by simonguinot — CHANGES_REQUESTED** [review]
Let's prevent this from being merged before @Mani-Sadhasivam or I review it.

File: boards/shields/npm1300_ek/npm1300_ek.overlay [hunk 4]
```diff
@@ -37,12 +39,14 @@
...
 				regulator-min-microvolt = <1000000>;
 				regulator-max-microvolt = <3300000>;
+				regulator-initial-mode = <NPM1300_LDSW_MODE_LDO>;
 				enable-gpios = <&npm1300_ek_gpio 2 GPIO_ACTIVE_LOW>;
 			};
 
 			npm1300_ek_ldo2: LDO2 {
 				regulator-min-microvolt = <1000000>;
 				regulator-max-microvolt = <3300000>;
+				regulator-initial-mode = <NPM1300_LDSW_MODE_LDSW>;
```
Comment (simonguinot) [inline boards/shields/npm1300_ek/npm1300_ek.overlay:49]: Is that related with LED support ?

**Review by simonguinot — CHANGES_REQUESTED** [review]
Hi @aasinclair,

The driver looks nice and straightforward. I just have a few comments. Please see them below.

File: dts/bindings/led/nordic,npm1300-led.yaml (hunk 3 above)
Comment (simonguinot) [reply dts/bindings/led/nordic,npm1300-led.yaml]: Nice.

File: drivers/led/led_npm1300.c [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+		uint8_t mode = config->mode[led];
+
+		if (mode <= NPM_LED_HOST) {
+			int ret = mfd_npm1300_reg_write(config->mfd, NPM_LED_BASE,
+							NPM_LED_OFFSET_MODE + led, mode);
+
+			if (ret < 0) {
+				return ret;
+			}
+		}
```
[the remark this answers is not in the document]
Comment (aasinclair) [reply drivers/led/led_npm1300.c:88]: Thinking more about this allowing the default / missing options is unnecessarily confusing.  
I'll make them all required and remove the `_OR`

File: boards/shields/npm1300_ek/npm1300_ek.overlay (hunk 4 above)
Comment (aasinclair) [reply boards/shields/npm1300_ek/npm1300_ek.overlay:49]: No it's not related, would you like this moved to a separate PR?

File: drivers/led/led_npm1300.c (hunk 5 above)
[the remark this answers is not in the document]
Comment (simonguinot) [reply drivers/led/led_npm1300.c:88]: I think it is a good idea.

File: boards/shields/npm1300_ek/npm1300_ek.overlay (hunk 4 above)
Comment (simonguinot) [reply boards/shields/npm1300_ek/npm1300_ek.overlay:49]: Yes please.

**Review by simonguinot — APPROVED** [review]
Thanks for this nice driver @aasinclair !

**Review by gmarull — APPROVED** [review]
